File: tools/demo/package.py
"""Package the live three-game demo for a static nginx server."""

import argparse
import gzip
import hashlib
import json
from pathlib import Path
import re
import shutil
import subprocess

ROOT = Path(__file__).resolve().parents[2]
GAMES = [('gota', 'gods_of_the_arena', 10),
         ('lvd', 'light_vs_dark', 2),
         ('cta', 'call_to_adventure', 4)]


def package(destination, release):
    """Copy runtime files and make every deployed URL independent of the repo."""
    if not re.fullmatch(r'[a-zA-Z0-9_-]+', release):
        raise ValueError('Release must contain only letters, numbers, _ or -.')
    destination.mkdir(parents=True, exist_ok=False)
    page = (ROOT / 'tools/demo/index.html').read_text()
    for game, directory, seats in GAMES:
        source = ROOT / 'examples' / directory
        bundle = source / 'emscripten'
        if 'configurePolyworldWebInputs' not in (bundle / f'{game}.js').read_text():
            raise ValueError(f'{game} needs a live -d:emscripten build.')
        target = destination / 'releases' / release / game
        target.mkdir(parents=True)
        for extension in ('html', 'js', 'wasm', 'data'):
            shutil.copyfile(bundle / f'{game}.{extension}',
                            target / f'{game}.{extension}')
        script = target / f'{game}.js'
        javascript = script.read_text()
        label = re.search(r"var PACKAGE_NAME = '([^']+)';", javascript)
        if not label or Path(label[1]).name != f'{game}.data':
            raise ValueError(f'{game} has an unknown asset package label.')
        javascript = javascript.replace(label[1], f'{game}.data')
        script.write_text(javascript)
        shutil.copyfile(source / 'players/base.bas', target / 'base.bas')
        old = (f'../../examples/{directory}/emscripten/{game}.html'
               f'?bot=../players/base.bas:{seats}')
        new = f'../releases/{release}/{game}/{game}.html?bot=base.bas:{seats}'
        if page.count(old) != 1:
            raise ValueError(f'The demo must contain exactly one {game} iframe.')
        page = page.replace(old, new)
    page = re.sub(r'    <!--.*?-->\n', '', page, flags=re.S)
    demo = destination / 'demo'
    demo.mkdir()
    (demo / 'index.html').write_text(page)
    files = {}
    for path in sorted(destination.rglob('*')):
        if not path.is_file():
            continue
        data = path.read_bytes()
        files[str(path.relative_to(destination))] = {
            'bytes': len(data), 'sha256': hashlib.sha256(data).hexdigest()
        }
        if path.suffix in ('.html', '.js', '.wasm', '.data', '.bas'):
            path.with_name(path.name + '.gz').write_bytes(
                gzip.compress(data, compresslevel=9, mtime=0))
    revision = subprocess.check_output(
        ['git', 'rev-parse', 'HEAD'], cwd=ROOT, text=True).strip()
    manifest = {
        'release': release,
        'source_revision': revision,
        'dependencies_sha256': hashlib.sha256(
            (ROOT / 'coworld/dependencies.lock').read_bytes()).hexdigest(),
        'files': files,
    }
    (destination / 'releases' / release / 'manifest.json').write_text(
        json.dumps(manifest, indent=2) + '\n')
    print(f'Packaged {len(files)} files for /polyworld/demo/ in {destination}')


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('destination', type=Path)
    parser.add_argument('--release', required=True)
    args = parser.parse_args()
    package(args.destination, args.release)

File: tools/demo/package.nim
import
  std/[algorithm, json, os, parseopt, strutils],
  crunchy, jsony, zippy,
  ../../coworld/tools/tooling

proc digest(data: string): string =
  ## Encodes the deployment manifest's SHA-256 checksum.
  for value in sha256(data):
    result.add value.toHex(2).toLowerAscii()

proc package(destination, release: string) =
  ## Copies standalone game files and writes deterministic deployment metadata.
  require(release.len > 0 and
    release.allCharsInSet({'a' .. 'z', 'A' .. 'Z', '0' .. '9', '_', '-'}),
    "Release must contain only letters, numbers, _ or -")
  require(not fileExists(destination) and not dirExists(destination) and
    not symlinkExists(destination), "Destination already exists: " & destination)
  createDir(destination)
  var page = readFile(Root / "tools/demo/index.html")
  for game in Games:
    let
      source = Root / "examples" / game.directory
      bundle = source / "emscripten"
      target = destination / "releases" / release / game.name
    var javascript = readFile(bundle / (game.name & ".js"))
    require("configurePolyworldWebInputs" in javascript,
      game.name & " needs a live -d:emscripten build")
    let label = between(javascript, "var PACKAGE_NAME = '", "';")
    require(label.extractFilename() == game.name & ".data",
      game.name & " has an unknown asset package label")
    javascript = javascript.replace(label, game.name & ".data")
    createDir(target)
    for extension in ["html", "wasm", "data"]:
      let filename = game.name & "." & extension
      copyFile(bundle / filename, target / filename)
    writeFile(target / (game.name & ".js"), javascript)
    copyFile(source / "players/base.bas", target / "base.bas")
    let
      oldUrl = "../../examples/" & game.directory & "/emscripten/" &
        game.name & ".html?bot=../players/base.bas:" & $game.seats
      newUrl = "../releases/" & release & "/" & game.name & "/" &
        game.name & ".html?bot=base.bas:" & $game.seats
    require(page.count(oldUrl) == 1,
      "The demo must contain exactly one " & game.name & " iframe")
    page = page.replace(oldUrl, newUrl)
  while true:
    let first = page.find("    <!--")
    if first < 0:
      break
    let last = page.find("-->\n", first)
    require(last >= 0, "Unterminated demo HTML comment")
    page.delete(first .. last + 3)
  createDir(destination / "demo")
  writeFile(destination / "demo/index.html", page)
  var paths: seq[string]
  for path in walkDirRec(destination):
    paths.add path
  paths.sort()
  let files = newJObject()
  for path in paths:
    let
      data = readFile(path)
      relative = relativePath(path, destination).replace('\\', '/')
    files[relative] = %*{"bytes": data.len, "sha256": digest(data)}
    if path.splitFile.ext in [".html", ".js", ".wasm", ".data", ".bas"]:
      writeFile(path & ".gz", compress(data, level = 9, dataFormat = dfGzip))
  let manifest = %*{
    "release": release,
    "source_revision": command(["git", "rev-parse", "HEAD"]),
    "dependencies_sha256": digest(readFile(Root / "coworld/dependencies.lock")),
    "files": files
  }
  writeFile(
    destination / "releases" / release / "manifest.json",
    manifest.toJson() & "\n"
  )
  echo "Packaged ", files.len, " files for /polyworld/demo/ in ", destination

proc main() =
  ## Parses the destination and required release identifier.
  var
    parser = initOptParser(commandLineParams(), longNoVal = @["help"])
    destination, release: string
    needsRelease = false
  for kind, key, value in parser.getopt():
    case kind
    of cmdArgument:
      if needsRelease:
        release = key
        needsRelease = false
      else:
        require(destination.len == 0, "Unexpected argument: " & key)
        destination = key
    of cmdLongOption:
      require(key == "release", "Unknown option: --" & key)
      release = value
      needsRelease = value.len == 0
    of cmdShortOption:
      raise newException(PolyworldToolsError, "Unknown option: -" & key)
    of cmdEnd:
      discard
  require(destination.len > 0 and release.len > 0 and not needsRelease,
    "Usage: nim r tools/demo/package.nim DESTINATION --release RELEASE")
  package(destination, release)

runTool(main)
